File: LeetCode/Study/ArrayAndString/22.MoveZeros.java
/**
 * 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
 * 请注意 ，必须在不复制数组的情况下原地对数组进行操作。
 */

class Solution {
    public void moveZeroes(int[] nums) {
        int lastNonZeroIndex = 0;

        // 先把所有非零元素按顺序移到前面
        for (int current = 0; current < nums.length; current++) {
            if (nums[current] != 0) {
                nums[lastNonZeroIndex] = nums[current];
                lastNonZeroIndex++;
            }
        }

        // 把剩下的位子补0
        for (int i = lastNonZeroIndex; i < nums.length; i++) {
            nums[i] = 0;
        }
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();

        int[][] testInputs = {
            {0, 1, 0, 3, 12},
            {0, 0, 1},
            {1, 2, 3},
            {0, 0, 0},
            {},
            {4, 0, 0, 1, 0, 3, 0, 5}
        };

        int[][] expectedOutputs = {
            {1, 3, 12, 0, 0},
            {1, 0, 0},
            {1, 2, 3},
            {0, 0, 0},
            {},
            {4, 1, 3, 5, 0, 0, 0, 0}
        };

        for (int i = 0; i < testInputs.length; i++) {
            int[] input = testInputs[i];
            sol.moveZeroes(input);

            System.out.println("Test case " + (i + 1) + ":");
            System.out.print("Result:   ");
            printArray(input);
            System.out.print("Expected: ");
            printArray(expectedOutputs[i]);
            System.out.println(compareArrays(input, expectedOutputs[i]) ? "✅ Pass" : "❌ Fail");
            System.out.println("----------------------------------------");
        }
    }

    private static void printArray(int[] arr) {
        System.out.print("[");
        for (int j = 0; j < arr.length; j++) {
            System.out.print(arr[j]);
            if (j < arr.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }

    private static boolean compareArrays(int[] a, int[] b) {
        if (a.length != b.length) return false;
        for (int i = 0; i < a.length; i++) {
            if (a[i] != b[i]) return false;
        }
        return true;
    }
}

File: LeetCode/Study/ArrayAndString/22.MoveZeros.py
"""
给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
请注意 ，必须在不复制数组的情况下原地对数组进行操作。
"""

from typing import List

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        last_non_zero = 0
        
        # 将非零元素移动到数组的前面
        for i in range(len(nums)):
            if nums[i] != 0:
                nums[last_non_zero] = nums[i]
                last_non_zero += 1

        # 将剩余的元素设置为 0
        for i in range(last_non_zero, len(nums)):
            nums[i] = 0

def test():
    sol = Solution()
    test_cases = [
        ([0, 1, 0, 3, 12], [1, 3, 12, 0, 0]),  # 移动 0 到末尾
        ([0, 0, 1], [1, 0, 0]),  # 多个连续的 0
        ([1, 2, 3], [1, 2, 3]),  # 没有 0
        ([0], [0]),  # 单个元素是 0
        ([1], [1]),  # 单个非零元素
    ]
    
    for i, (nums, expected) in enumerate(test_cases, 1):
        sol.moveZeroes(nums)
        print(f"Test case {i}:")
        print(f"Input: {nums}")
        print(f"Expected: {expected}, Result: {nums}")
        print("✅ Pass" if nums == expected else "❌ Fail")
        print("-" * 40)

if __name__ == "__main__":
    test()